lib/datadog/profiling/ext/dir_monkey_patches.rb

```diff
@@ -0,0 +1,223 @@
+            super
+          end
+        end
+        ruby2_keywords :foreach if respond_to?(:ruby2_keywords, true)
+
+        def glob(*args, &block)
+          if block
+            begin
+              # <-- Begin critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super do |entry_name|
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+                # <-- We're safe now while running customer code
+                yield entry_name
+                # <-- We'll go back to the Dir internals, critical region again
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              end
+            ensure
+              # <-- End critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+            end
+          else
+            begin
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super
+            ensure
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+            end
+          end
+        end
+        ruby2_keywords :glob if respond_to?(:ruby2_keywords, true)
+
+        def home(*args, &block)
+          Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+          super
+        ensure
+          Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+        end
+        ruby2_keywords :home if respond_to?(:ruby2_keywords, true)
+      end
+
+      # Monkey patches for Dir. See DirMonkeyPatches above for more details.
+      module DirInstanceMonkeyPatches
+        def each(*args, &block)
+          if block
+            begin
+              # <-- Begin critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super do |entry_name|
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+                # <-- We're safe now while running customer code
+                yield entry_name
+                # <-- We'll go back to the Dir internals, critical region again
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              end
+            ensure
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions # <-- End critical region
+            end
+          else
+            # This returns an enumerator. We don't want/need to intercede here, the enumerator will eventually call the
+            # other branch once it gets going.
+            super
+          end
+        end
+        ruby2_keywords :each if respond_to?(:ruby2_keywords, true)
+
+        unless RUBY_VERSION.start_with?('2.5.') # This is Ruby 2.6+
```

Comment:
The `RUBY_VERSION > '2.5'` pattern is very common, and will work if someone alters gemspec etc. to run on Ruby 2.4 for some reason, I would use it on account of its greater reliability.

Reply:
The thing is -- there's more methods that are not there on 2.4, 2.3 etc.

So you'll need to do other changes anyway, so I'm not sure it's worth trying to half-support older Rubies, and for any weird testing, you can disable this whole thing anyway...

Or do you feel strongly this is not a good idea?

Reply:
Not at all, please disregard.